### PR TITLE
Read metadata info protobuf format

### PR DESCRIPTION
**Pull request by NRHelmi**
This PR adds support to protobuf metadata
```
cd examples
go run execute/main.go -e ENGINE -d DB -c "def output = 1"
```

```
{
    "GotCompleteResult": true,
    "Transaction": {
        "id": "206c49ca-bdc7-c7c8-8edd-ad49a211455b",
        "state": "COMPLETED"
    },
    "Results": [
        {
            "RelationID": "/:output/Int64",
            "Table": [
                1
            ]
        }
    ],
    "Metadata": {
        "relations": [
            {
                "relation_id": {
                    "arguments": [
                        {
                            "tag": 3,
                            "constant_type": {
                                "rel_type": {
                                    "tag": 1,
                                    "primitive_type": 17
                                },
                                "value": {
                                    "arguments": [
                                        {
                                            "tag": 17,
                                            "Value": {
                                                "StringVal": "b3V0cHV0"
                                            }
                                        }
                                    ]
                                }
                            }
                        },
                        {
                            "tag": 1,
                            "primitive_type": 2
                        }
                    ]
                },
                "file_name": "0.arrow"
            }
        ]
    },
    "Problems": []
}

```

### CODE REVIEW

**Comment by larf311** [comment]
What is a good way to test this? The example just prints nothing. I'm querying a transaction that was done on a an engine created yesterday.

**Comment by NRHelmi** [comment]
> What is a good way to test this? The example just prints nothing. I'm querying a transaction that was done on a an engine created yesterday.

I created a PR description but missed the update button :sweat_smile:, should be updated now :smile:

**Comment by larf311** [comment]
Ha ha, my comment was actually for the C# sdk. I got my browser tabs mixed up.

File: rai/client.go [hunk 1]
```diff
@@ -318,6 +325,11 @@ func unmarshal(rsp *http.Response, result interface{}) error {
 		}
 
 		return errors.Errorf("unhandled unmarshal type %T", result)
+	case generated.MetadataInfo:
+		dstValues := reflect.ValueOf(result).Elem()
```
Comment (bachdavi) [inline rai/client.go:329]: Is `dstValues` not already assigned in line 281? I'm not sure about Go scoping, but I guess it doesn't matter here as we are mutating the `result` argument...

File: rai/client.go [hunk 2]
```diff
@@ -342,7 +354,7 @@ func readArrowFiles(files []TransactionAsyncFile) ([]ArrowRelation, error) {
 					data, _ := rec.Column(i).MarshalJSON()
 					var values []interface{}
 					json.Unmarshal(data, &values)
-					out = append(out, ArrowRelation{rec.ColumnName(i), values})
+					out = append(out, ArrowRelation{file.Name, values})
```
Comment (bachdavi) [inline rai/client.go:357]: This is unrelated to the current PR, but I'm wondering what the `json.Unmarshal` is doing here in `readArrowFiles`? :) 
The change is unrelated as well, right?

File: rai/client.go [hunk 3]
```diff
@@ -426,19 +438,28 @@ func parseHttpResponse(rsp *http.Response) (interface{}, error) {
...
 	mediaType, params, _ := mime.ParseMediaType(rsp.Header.Get("Content-Type"))
+
 	if mediaType == "application/json" {
 		return data, nil
 	} else if mediaType == "multipart/form-data" {
 		boundary := params["boundary"]
 		return parseMultipartResponse(data, boundary)
+	} else if mediaType == "application/x-protobuf" {
+		var metadataInfo generated.MetadataInfo
+		err := proto.Unmarshal(data, &metadataInfo)
```
Comment (bachdavi) [inline rai/client.go:449]: Again not sure I understand Go ^^ but we seem not to be using our function `unmarshal` that is defined above.

Comment (bachdavi) [reply rai/client.go:449]: Ah, I see, `unmarshal` is called in `request`. But before that we use this method `parseHTTPResponse` and in here we actually unmarshal the Proto and above in `unmarshal` we simply pass it through :)

**Review by bachdavi — APPROVED** [review]
LGTM 🎉 
Just some more minor comments :)

I would request two things:
1. Can you add a `metadata.pb.txt` version of the test file just so that its easier to understand what it is? There are different ways to generate these, e.g. `MessageToString` https://googleapis.dev/python/protobuf/latest/google/protobuf/text_format.html

2. Can you bump the version and add a CHANGELOG entry? This is a breaking change as the JSON metadata is not returned anymore

File: rai/client.go (hunk 2 above)
Comment (NRHelmi) [reply rai/client.go:357]: the `data, _ := rec.Column(i).MarshalJSON()` returns a `byte array` one way to convert it to an array of values is to use 
```
var values []interface{}
json.Unmarshall(data, &values)
```
Yep the change is unrelated and it's mainly added to get aligned with `js-sdk` it sets the relation id to the rel path.

File: rai/client.go (hunk 1 above)
Comment (NRHelmi) [reply rai/client.go:329]: yep it's already assigned, this line is redundant, cleaning it up :+1:

**Comment by NRHelmi** [comment]
> LGTM tada Just some more minor comments :)
> 
> I would request two things:
> 
> 1. Can you add a `metadata.pb.txt` version of the test file just so that its easier to understand what it is? There are different ways to generate these, e.g. `MessageToString` https://googleapis.dev/python/protobuf/latest/google/protobuf/text_format.html
> 2. Can you bump the version and add a CHANGELOG entry? This is a breaking change as the JSON metadata is not returned anymore

this might be useful for future debugging 
```
protoc -I ~/workspace/rai-sdk-go/protos/proto/ --decode=relationalai.protocol.MetadataInfo ~/workspace/rai-sdk-go/protos/proto/message.proto < rai/metadata.pb > rai/metadata.pb.txt

```

**Comment by NRHelmi** [comment]
If no other updates are required I'll merge this by the end of the day :pray:

File: CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,11 @@
 # Changelog
 
+## v0.5.0-alpha
+* Deprecated metadata json format.
+* Removed `TransactionAsyncMetadataResponse` model.
+* Added support to metadata protobuf format.
+* `GetTransactionMetadata` returns protobuf metadata.
+
```
Comment (NHDaly) [inline CHANGELOG.md:8]: Oh, I didn't realize that there were version numbers already in the CHANGELOG. ... 🤔 

@vilterp  and I created two new versions while you were out - versions 0.2.0 and 0.2.1:
https://github.com/RelationalAI/rai-sdk-go/tags

But maybe those should have already been 0.4.0?

How are you currently indicating the version numbers other than in this changelog file? It seems that the Go package manager finds version numbers through github tags, which is why we added those. But maybe we were missing some other existing mechanism?

Comment (NHDaly) [reply CHANGELOG.md:8]: I'm not sure what to do about the version numbers we already tagged... 😅 Maybe it's fine to delete them and set the tags back to match up with version numbers in this file? Or maybe we just leave them and then skip straight to `v0.5.0-alpha`?

**Review by NHDaly — APPROVED** [review]
🎉 amazing, thanks @NRHelmi! Agreed that this LGTM! :)

File: CHANGELOG.md (hunk 4 above)
Comment (NRHelmi) [reply CHANGELOG.md:8]: I was planning to add a workflow to index this package to `pkg.go.dev` but didn't implement that :sweat_smile:, happy to hear that the Go package manager finds version numbers also through github tags :smile:

Comment (NHDaly) [reply CHANGELOG.md:8]: So how should we tag a new release, then? Maybe we can discuss tomorrow? Thanks helmi!

**Pull request merged**